quic/s2n-quic-transport/src/ack/ack_ranges.rs

```diff
@@ -0,0 +1,259 @@
+        let range_0 = PacketNumberRange::new(pn_a, pn_a);
+        let range_1 = PacketNumberRange::new(pn_b, pn_c);
+        let range_2 = PacketNumberRange::new(pn_d, pn_f);
+        let range_3 = PacketNumberRange::new(pn_g, pn_h);
+        let range_4 = PacketNumberRange::new(pn_h, pn_i);
+        let range_0_1 = PacketNumberRange::new(pn_a, pn_c);
+        let range_1_2 = PacketNumberRange::new(pn_c, pn_e);
+
+        assert!(ack_ranges.insert_packet_number_range(range_1).is_ok());
+        assert!(ack_ranges.insert_packet_number_range(range_2).is_ok());
+        assert!(ack_ranges.insert_packet_number_range(range_3).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 3);
+
+        for pn in range_1 {
+            assert!(ack_ranges.contains(&pn));
+        }
+        for pn in range_2 {
+            assert!(ack_ranges.contains(&pn));
+        }
+        for pn in range_3 {
+            assert!(ack_ranges.contains(&pn));
+        }
+
+        // merge ranges 1 and 2
+        assert!(ack_ranges.insert_packet_number_range(range_1_2).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 2);
+
+        // insert range 0 at low end
+        assert!(ack_ranges.insert_packet_number_range(range_0).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 3);
+
+        // merge range 0_1 at low end
+        assert!(ack_ranges.insert_packet_number_range(range_0_1).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 2);
+
+        // merge new range at high end
+        assert!(ack_ranges.insert_packet_number_range(range_4).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 2);
+    }
+
+    #[test]
+    fn large_range_test() {
+        let pn_a = PacketNumberSpace::ApplicationData.new_packet_number(VarInt::from_u32(1));
+        let pn_b = PacketNumberSpace::ApplicationData
+            .new_packet_number(VarInt::new(varint::MAX_VARINT_VALUE).unwrap());
+        let mut ack_ranges = AckRanges::new(3);
+
+        let range_1 = PacketNumberRange::new(pn_a, pn_b);
+
+        assert!(ack_ranges.insert_packet_number_range(range_1).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 1);
+    }
+
+    #[test]
+    fn size_of_snapshots() {
+        use core::mem::size_of;
+        use insta::assert_debug_snapshot;
+
+        assert_debug_snapshot!("AckRanges", size_of::<AckRanges>());
+    }
+
+    #[test]
+    fn insert_packet_number_range_fuzz() {
+        let mut ack_ranges = AckRanges::new(10);
+        check!()
+            .with_type::<(u32, u32)>()
+            .map(|(a, b)| (a.min(b), a.max(b))) // ensure valid range
+            .for_each(|(a, b)| {
+                let pn_a = PacketNumberSpace::Initial.new_packet_number(VarInt::from_u32(*a));
+                let pn_b = PacketNumberSpace::Initial.new_packet_number(VarInt::from_u32(*b));
+                let range_1 = PacketNumberRange::new(pn_a, pn_b);
+
+                assert!(ack_ranges.insert_packet_number_range(range_1).is_ok());
```

Comment:
I don't think we can say that this is ok. We're _mostly_ just testing that we don't panic anywhere. One thing you could do is if we get an error, the range should be lower than the minimum.

Reply:
What I actually wanted `with_type<Vec<(u32, u32)>>` and create distince ranges in the interval set. But that seems too complicated for the fuzzer and doesnt product interesting test cases. This is because adding overlapping ranges collapses into a single range and reduces into a less interesting case.

 Pushing a new change that will move `AckRanges::new(1);` inside the loop(recreate a new interval set) and verify that `insert.is_ok()` since we should have one spot. **This should verify large range inserts are safe which is primarily what we wanted.**